man/man6/wreversi.6
.TH WREVERSI 6 "Version 1, Release 4" "W Window System" "W PROGRAMS"
.SH NAME
wreversi \- Reversi / Othello board game client for W
.SH SYNOPSIS
.nf
wreversi
.fi
.SH DESCRIPTION
\fIWRversi\fP is a W version of Reversi / Othello board game clients
for the networked turn based, two player game framework by Eero Tamminen.
.SH STARTING GAME
Before starting a game one has to contact another game client.  This
happens either by setting up a server and accepting a connection or by
connecting yourself to another game server. You can do this from
the configuration dialog.
.PP
With curses version one needs to use '-c' option to specify the address
into which to connect (e.g.  'creversi -c saturn.outer.space.org').  If
this option isn't used, the game will itself setup a server.
.PP
After connecting another player, you can start the game and begin to
play.  Note that continuing (for clients that support board editing
before play) asks confirmation from the other player before proceeding.
.SH REVERSI / OTHELLO RULES
Players position one stone onto the board at the time.  The opponent's
pieces that are between (horizontally, vertically and diagonally) the
new piece and the earlier pieces of the same color, will be converted to
the new color.  Only moves which convert opponent's pieces are legal.
Game ends when both players pass.  The player with most stones on the
board wins.
.SH SEE ALSO
.IR creversi (6),
.IR reversid (6)
.SH LAST CHANGED
09/2009 by Eero Tamminen


wt/CHANGES
++oddie 9/2000:
- fixed listbox/filesel resizing & added minimum size setopt.

++bsittler 7/2000:
- Fixed vt color/reverse and attribute resetting
- Add HTTP/1.1 'Host' header to Wetscape requests
- Other minor changes (netgames, wvt, waudio)

++jcarlson 7/2000:
- WTINYSCREEN changes

++eero, 5/1999:
- Fixed key handling for special keys in Wt/getstring.c & wedit/wedit.c.

++eero, 1/1999:
- Added support for new W special keys so that now on W platforms which
implement special key mappings, you don't need to use Emacs shortcuts
with text widgets (or remap your console cursor keys)...


------ notes for version released with W R1PL0 ----------

ADDITIONAL WIDGETS

- Arrow, a simple button with an arrow drawn into it.

- Dial, a dial widget where you can position a needle into desired
  angle.  This isn't in the W toolkit library itself as then you would
  need to link math library to all of your Wt programs.  You'll find it
  in the warzone game tar-ball.

- Edittext, mainly intended for simple one page stuff, but as this
  spaghetti can do some additional stuff like paragraph formatting,
  there's a real abuse built on this, the WEdit.  As edittext widget is
  large and buggy, it's in the WEdit directory instead of the toolkit.

- Menu, a strip of buttons from which popups pop-up.  Both the popups
  and menus use a structure defined in Wt.h instead of strings that Kay
  was using.  The reason is that I want callbacks to follow
  automatically hierarchically opened popups.

- Popup, hierarchical popups.  Can include subpopups, images, lines and
  in addition items can also be disabled and checked.

- Range, a scrollbar composite with which you can select a value between
  given limits.

- Select, a getstring widget coupled with an arrow openable listbox from
  which items can be selected into a getstring widget.


OTHER CHANGES

- vt can now do everything wterm  can do + user can rebind the colors
  in wtrc.

- Added '.wtconfig' file and made makefiles use the W compilation
  configuration file '.config' in parent directory.

- wt_do_event() doesn't call wt_dispatch_event() anymore with
  EVENT_GADGET/GADGET_EXIT event as those might not have valid window
  pointer.  If you want to check for abnormal (unboud) exit, check the
  wt_return value.  GADGET_EXIT returns a negative number and
  GADGET_CLOSE (in shell widget) a positive number now.

- Added xterm mouse events, reverse mode, fd/bgcolor setting + improved
  double/triple click timing in VT widget.

- Aliases a couple new tags into others in HTML widget and added
  HTML header (ignored) and TITLE parsing.

- Widget event processing functions return now unrecognized or
  non-processed WEVENTs or NULL instead of long.  Assumption is that if
  function returns WEVENT, the widget whose action it was, is still
  around.  If in doubt, use NULL.  Unrecognized key events should be
  returned for focus change though.  (actually it could still have been
  an int, but returning a WEVENT offers a possibility for event mapping
  widgets.)

- Added wt_global structure containing Wt configuration variables and
  made widgets to use it.  wt_init() will now read these variables from
  ~/.wtrc + a list of (variable, value) pairs.  Widget specific
  initialization code in their init() functions may read these with the
  wt_variable() function.

- HTML widget fonts can be configured by above method.

- Added wt_loadfont() function which will use the user specified Wt
  default if widget doesn't specify some font parameter.  It also exits
  program if Wt default fonts can't be loaded, because otherwise
  unavailable fonts may chrash naive programs.

- Changed widgets to reflect the new w_loadfont() syntax (NULL/zero
  values tell Wlib/server to use server defaults).

- Added new WT_FONTSIZE define.  The earlier WT_FONT setopt sets now the
  font *family* name, new one sets the font size.

- Getopting WT_FONT returns WFONT* pointer also for above reason.

- Removed 'fontname' members from widget structures.  Needed information
  and more can be gotten from the 'font' member if needed.

- Fileselector has now separate lists for directory and file names.
  Both of course have an autolocator (type & return in addition to point
  & click).

- Wt_keys.h defines values for some special keys like cursors etc.  Half
  of the defines are Emacs style and half what the special keys itself
  give or are in use on other platforms.

- Changed Wt option defines.  There were a couple overlapping ones.


------- Kay's original notes: ----------------------------------------------

- added some missing close() and truncate the file to zero length before
  writing a new selection in selection.c.

- better selection handling in VT widget.

- made everything resizeable.

- applied patches by eero and fixed some more bugs myself.

- implemented a simple clipboard and changed getstring and vt widget to use it.

- fixed several serious bugs in the html widget.

- wrote VT widget.

- added an option to the icon editor to sync edit field with display
  field (simon).

- added missing special chars to html widget (simon).

- added code to the fileselector to set a default file name.

- lots of changes by eero (button.c, fileselect.c, listbox.c, getstring.c), some
  fixes to his changes by me.

- added missing wt_ungetfocus() to listbox and getstring.

- wrote Tk-like packer widget.

- new fileselect widget and testprogram by eero. Added better filename
  matching (eeros version was a joke, really).

- getstring and listbox updates by eero.

- changed the scrollbar widget's user interface to become more user friendly.
  See scrollbar.doc for details on how to use it. Also changed viewport and
  listbox to use the new interface.

- added listbox_focus(). Otherwise listbox cannot receive keyboard focus...

- big bugfix: when wt_delete() is called on a widget all pending timers must
  be canceled! (affected list, iconedit and getstring).


games/wreversi.c
/*
 * wreversi, a part of the W Window System
 *
 * Copyright (C) 1994-1997 by Torsten Scherer
 *
 * This package is free software; you can redistribute it and/or modify it
 * under the terms specified in the docs/COPYRIGHTS file coming with this
 * package.
 *
 * -- this is my old `wreversi' program I've written over a year ago, but
 * never really got it ready.  in fact it only served as a test platform for
 * the pcircle call.  the algorithm is surely not very clever, but then I
 * wrote it out of scratch without any further knowledge of other ways of
 * doing it, i.e. I did *not* look into some other program and copied that.
 * :)
 *
 * CHANGES
 * ++TeSche, 02/96:
 * - adapted for W1R3, but still needs an awful lot of fixes I guess
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <Wlib.h>

#define	BWIDTH	8
#define	BHEIGHT	8

#define	WIDTH		(BWIDTH*20)
#define	HEIGHT		(BHEIGHT*20)
#define	MAXLOOSE	(- BWIDTH * BHEIGHT)

typedef	struct {
  short field[BWIDTH][BHEIGHT];
} BOARD;

char icon[] = ".o*";
short maxdepth;


/*
 * the graphic parts...
 */

static WWIN *glob_win;
static short glob_z1, glob_z2, glob_yoffset;
static WFONT *glob_font;


/*
 *
 */

static WWIN *openwindow(short width, short height)
{
  WWIN *ret;

  if (!w_init()) {
    return NULL;
  }

  if (!(glob_font = w_loadfont("fixed", 13, 0))) {
    return NULL;
  }

  glob_z1 = 1;
  glob_z2 = 3 + glob_font->height;
  glob_yoffset = 2 * (glob_font->height + 2);

  if (!(ret = w_create(width, height + glob_yoffset,
		       W_MOVE | W_TITLE | W_CLOSE | EV_MOUSE))) {
    return NULL;
  }

  w_settitle(ret, " wreversi ");
  w_setfont(ret, glob_font);

  if (w_open(ret, UNDEF, UNDEF) < 0) {
    w_delete(ret);
    return NULL;
  }

  return ret;
}


static void showboard(BOARD *bp)
{
  int x, y;

  for (y=0; y<BHEIGHT; y++) {
    for (x=0; x<BWIDTH; x++) {
      if (bp->field[x][y]) {
	switch (bp->field[x][y]) {
	  case 1:
	    w_setmode(glob_win, M_CLEAR);
	    break;
	  case 2:
	    w_setmode(glob_win, M_DRAW);
	  }
	w_pcircle(glob_win, x*20+10, glob_yoffset+y*20+10, 8);
      }
    }
  }
}


/*
 * the reversi functions...
 */

static long checked;
static char buf[80];

static void init(BOARD *bp)
{
  int i, j;

  for (i=0; i<BWIDTH; i++) {
    for (j=0; j<BHEIGHT; j++) {
      bp->field[i][j] = 0;
    }
  }

  bp->field[BWIDTH/2-1][BHEIGHT/2-1] = 1;
  bp->field[BWIDTH/2][BHEIGHT/2] = 1;
  bp->field[BWIDTH/2][BHEIGHT/2-1] = 2;
  bp->field[BWIDTH/2-1][BHEIGHT/2] = 2;
}


static short flipdir(BOARD *fp,
		     short party,
		     short x, short y,
		     short dx, short dy)
{
  int lx, ly, count, end;

  /* first check if this would give legal flips */

  lx = x;
  ly = y;
  end = 0;
  count = 0;

  while (!end) {

    lx += dx;
    ly += dy;

    if ((lx<0) || (lx>=BWIDTH) || (ly<0) || (ly>=BHEIGHT)) {
      end = 1;
      count = 0;
    } else if (fp->field[lx][ly] == 0) {
      end = 1;
      count = 0;
    } else if (fp->field[lx][ly] == party) {
      end = 1;
    } else {
      count++;
    }
  }

  /* then do the flips */

  if (count) {

    lx = x;
    ly = y;
    end = 0;

    while (!end) {

      lx += dx;
      ly += dy;

      if (fp->field[lx][ly] == party) {
	end = 1;
      } else {
	fp->field[lx][ly] = party;
      }
    }
  }

  return count;
}


static short flips(BOARD *bp, short party, short x, short y)
{
  short count;

  if (!(++checked % 2500)) {
    w_setmode(glob_win, M_CLEAR);
    w_pbox(glob_win, 0, 0, WIDTH, glob_yoffset);
    sprintf(buf, "%li checked...", checked);
    w_printstring(glob_win, 1, glob_z2, buf);
    w_flush();
  }

  count  = flipdir(bp, party, x, y, -1, -1);
  count += flipdir(bp, party, x, y, -1,  0);
  count += flipdir(bp, party, x, y, -1, +1);
  count += flipdir(bp, party, x, y,  0, -1);
  count += flipdir(bp, party, x, y,  0, +1);
  count += flipdir(bp, party, x, y, +1, -1);
  count += flipdir(bp, party, x, y, +1,  0);
  count += flipdir(bp, party, x, y, +1, +1);

  if (count) {
    bp->field[x][y] = party;
  }

  return count;
}


static short best(BOARD *bp, short party, short depth,
		  short *zugx, short *zugy)
{
  short	i, j, max, new, mx = 0, my = 0;
  BOARD	tmp;

  /* the termination condition */

  if (depth > maxdepth) {
    return 0;
  }

  max = MAXLOOSE;
  for (i=0; i<BWIDTH; i++) {
    for (j=0; j<BHEIGHT; j++) {
      if (!bp->field[i][j]) {
	tmp = *bp;
	tmp.field[i][j] = party;

	/* is this step legal? */
	if ((new = flips(&tmp, party, i, j)) > 0) {

	  if (depth & 1) {
	    new += best(&tmp, 3-party, depth+1, zugx, zugy);
	  } else {
	    new -= best(&tmp, 3-party, depth+1, zugx, zugy);
	  }

	  if (new > max) {
	    mx = i;
	    my = j;
	    max = new;
	  }
	}
      }
    }
  }

  if (max != MAXLOOSE) {
    bp->field[mx][my] = party;
    flips(bp, party, mx, my);
    *zugx = mx;
    *zugy = my;
  }

  return max;
}


/*
 *
 */

int main(int argc, char *argv[])
{
  BOARD board, tmp;
  short party, i, j, weiss, schwarz;
  short zugx, zugy;
  short flipped, user = 1;
  WEVENT *wevent;

  maxdepth = 1;
  if (argc == 2) {
    if ((maxdepth = atoi(argv[1])) < 1) {
      maxdepth = 1;
    }
  }

  if (!(glob_win = openwindow(WIDTH, HEIGHT))) {
    fprintf(stderr, "error: wreversi can't open a window\n");
    return -1;
  }

  w_dpbox(glob_win, 0, glob_yoffset, WIDTH, HEIGHT);
  w_setmode(glob_win, M_DRAW);
  for (j=0; j<HEIGHT; j+=20) {
    for (i=0; i<WIDTH; i+=20) {
      w_box(glob_win, i, glob_yoffset+j, 20, 20);
    }
  }

  init(&board);
  showboard(&board);

  party = 1;

  while (42) {

    /* just check if the next party could move at all */

    checked = 0;
    tmp = board;

    if (best(&tmp, party, 1, &zugx, &zugy) == MAXLOOSE) {

      w_setmode(glob_win, M_CLEAR);
      w_pbox(glob_win, 0, 0, WIDTH, glob_yoffset);

      sprintf(buf, "`%c' can't move!", icon[party]);
      w_printstring(glob_win, 1, glob_z1, buf);

      weiss = 0;
      schwarz = 0;
      for (i=0; i<BWIDTH; i++) {
	for (j=0; j<BHEIGHT; j++) {
	  if (board.field[i][j] == 1) {
	    weiss++;
	  } else if (board.field[i][j] == 2) {
	    schwarz++;
	  }
	}
      }
      sprintf(buf, "result: o=%i, *=%i", weiss, schwarz);
      w_printstring(glob_win, 1, glob_z2, buf);

w_flush();
sleep(5);

      w_queryevent(NULL, NULL, NULL, -1);

      w_close(glob_win);
      w_delete(glob_win);

      return -1;
    }

    if (party == user) {

      w_setmode(glob_win, M_CLEAR);
      w_pbox(glob_win, 0, 0, WIDTH, glob_yoffset);
      sprintf(buf, "it's your turn...");
      w_printstring(glob_win, 1, glob_z1, buf);

      while (w_queryevent(NULL, NULL, NULL, 0))
	;

      flipped = 0;
      do {
	if ((wevent = w_queryevent(NULL, NULL, NULL, -1))) {
	  switch (wevent->type) {
	    case EVENT_GADGET:
	      if ((wevent->key == GADGET_EXIT) ||
		  (wevent->key == GADGET_CLOSE)) {
		return -1;
	      }
	      break;
	    case EVENT_MPRESS:
	      if (wevent->y > glob_yoffset) {
		tmp = board;
		zugx = wevent->x / 20;
		zugy = (wevent->y - glob_yoffset) / 20;
		flipped = flips(&tmp, party, zugx, zugy);
	      }
	  }
	}
      } while (!flipped);

    } else {

      sprintf(buf, "it's my turn...");
      w_printstring(glob_win, 1, glob_z1, buf);
      w_flush();

      checked = 0;
      tmp = board;
      best(&tmp, party, 1, &zugx, &zugy);

      w_setmode(glob_win, M_CLEAR);
      w_pbox(glob_win, 0, 0, WIDTH, glob_yoffset);
      sprintf(buf, "%li checked...", checked);
      w_printstring(glob_win, 1, glob_z2, buf);
    }

    for (i=0; i<10; i++) {
      if (i & 1) {
	w_setmode(glob_win, M_DRAW);
      } else {
	w_setmode(glob_win, M_CLEAR);
      }
      w_pcircle(glob_win, zugx*20+10, glob_yoffset+zugy*20+10, 8);
      w_flush();
      usleep(50000);
    }

    board = tmp;
    showboard(&board);

    party = 3 - party;
  }
}
